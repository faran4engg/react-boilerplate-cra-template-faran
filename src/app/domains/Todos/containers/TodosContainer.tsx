import { FC, memo } from 'react';

import { getAllTodos } from './queries';

import { OwnProps, RenderProps } from './types';
import { useQuery } from 'react-query';

const TodosContainer: FC<OwnProps & RenderProps> = ({ children }) => {
  const { isLoading, error, data } = useQuery('todos', getAllTodos, {
    refetchOnWindowFocus: false,
    suspense: true,
  });

  if (isLoading) return <h1>Loading...'</h1>;

  if (error) return <p>An error has occurred'</p>;

  return (
    <div className="text-gray-600 dark:text-gray-400">
      {children({
        isLoading: false,
        todos: data,
      })}
    </div>
  );
};

export default memo(TodosContainer);
