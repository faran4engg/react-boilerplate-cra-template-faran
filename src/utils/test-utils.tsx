export const asMock = (func: Function) => {
  if (!jest.isMockFunction(func)) {
    console.error('[asMock]: error');
  }
  return func as jest.Mock;
};
export const createRenderPropMock = (mock: jest.Mock, props?: {}) => {
  mock.mockImplementation(({ children }) => children(props));
};

export const createSafeRenderPropMock = (component: Function, props?: {}) => {
  createRenderPropMock(asMock(component), props);
};
