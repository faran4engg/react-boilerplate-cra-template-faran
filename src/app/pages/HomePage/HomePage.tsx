import { Helmet } from 'react-helmet-async';
import { TodoCard } from 'app/domains/Todos/components/todo-card';
import TodosContainer from 'app/domains/Todos/containers';

const HomePage = () => {
  return (
    <>
      <Helmet>
        <title>Home Page</title>
        <meta name="description" content="A Boilerplate application homepage" />
      </Helmet>
      <TodosContainer>{props => <TodoCard {...props} />}</TodosContainer>
    </>
  );
};

export default HomePage;
