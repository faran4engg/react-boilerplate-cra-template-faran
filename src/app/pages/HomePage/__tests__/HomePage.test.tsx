import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';

import { createSafeRenderPropMock } from 'utils/test-utils';
import TodosContainer from 'app/domains/Todos/containers';

import HomePage from '..';

jest.mock('app/domains/Todos/containers', () => jest.fn());

describe('<HomePage />', () => {
  it('should render and match the snapshot', () => {
    const renderer = createRenderer();
    renderer.render(<HomePage />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('should render with render prop', () => {
    const renderProp = {
      isLoading: false,
      todos: [],
    };
    createSafeRenderPropMock(TodosContainer, renderProp);

    const renderer = createRenderer();
    renderer.render(<HomePage />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
