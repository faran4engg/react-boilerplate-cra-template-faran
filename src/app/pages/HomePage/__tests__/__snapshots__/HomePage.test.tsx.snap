// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<React.Fragment>
  <Helmet
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Home Page
    </title>
    <meta
      content="A Boilerplate application homepage"
      name="description"
    />
  </Helmet>
  <mockConstructor>
    [Function]
  </mockConstructor>
</React.Fragment>
`;

exports[`<HomePage /> should render with render prop 1`] = `
<React.Fragment>
  <Helmet
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Home Page
    </title>
    <meta
      content="A Boilerplate application homepage"
      name="description"
    />
  </Helmet>
  <mockConstructor>
    [Function]
  </mockConstructor>
</React.Fragment>
`;
